include/sys/cbprintf_cxx.h

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_CBPRINTF_CXX_H_
+#define ZEPHYR_INCLUDE_SYS_CBPRINTF_CXX_H_
+#ifdef __cplusplus
+
+/* C++ version for detecting a pointer to a string. */
+static inline int z_cbprintf_cxx_is_pchar(char *)
+{
+	return 1;
+}
+
+static inline int z_cbprintf_cxx_is_pchar(const char *)
+{
+	return 1;
+}
+
+static inline int z_cbprintf_cxx_is_pchar(volatile char *)
+{
+	return 1;
+}
+
+static inline int z_cbprintf_cxx_is_pchar(const volatile char *)
+{
+	return 1;
+}
+
+static inline int z_cbprintf_cxx_is_pchar(wchar_t *)
+{
+	return 1;
+}
+
+static inline int z_cbprintf_cxx_is_pchar(const wchar_t *)
+{
+	return 1;
+}
+
+static inline int z_cbprintf_cxx_is_pchar(volatile wchar_t *)
+{
+	return 1;
+}
+
+static inline int z_cbprintf_cxx_is_pchar(const volatile wchar_t *)
+{
+	return 1;
+}
+
+template < typename T >
+static inline int z_cbprintf_cxx_is_pchar(T arg)
+{
+	return 0;
+}
+
+/* C++ version for calculating argument size. */
+static inline size_t z_cbprintf_cxx_arg_size(float f)
+{
+	return sizeof(double);
+}
+
+template < typename T >
+static inline size_t z_cbprintf_cxx_arg_size(T arg)
+{
+	return sizeof(arg + 0);
+}
+
+/* C++ version for storing arguments. */
+static inline void z_cbprintf_cxx_store_arg(uint8_t *dst, float arg)
+{
+	double d = (double)arg;
+
+	z_cbprintf_wcpy((int *)dst, (int *)&d, sizeof(d) / sizeof(int));
+}
+
+template < typename T >
+static inline void z_cbprintf_cxx_store_arg(uint8_t *dst, T arg)
+{
+	size_t wlen = z_cbprintf_cxx_arg_size(arg) / sizeof(int);
+
+	z_cbprintf_wcpy((int *)dst, (int *)&arg, wlen);
+}
+
+/* C++ version for long double detection. */
+static inline bool z_cbprintf_cxx_is_longdouble(long double arg)
```

Comment:
Why does this return `bool` when the equivalent for `pchar` returns `int`?

Reply:
is_pchar is used in FOREACH loop to calculate number of char pointers in the argument list. This one is used for assert. I can change that to int for consistency.

Reply:
I think that makes sense, since the C version appears to evaluate to 0 or 1.

Reply:
done